src/test/ReactShallowRendererFiber.js

```diff
@@ -0,0 +1,124 @@
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactShallowRendererFiber
+ * @flow
+ */
+
+'use strict';
+
+const React = require('React');
+const emptyObject = require('emptyObject');
+const ReactFiberReconciler = require('ReactFiberReconciler');
+
+type Props = {
+  children : Array<any>,
+};
+
+type Container = {|
+  children : Array<Instance | TextInstance | null>,
+  context : Object,
+|};
+
+type Instance = {|
+  type : string,
+  props : Props,
+  children : Array<Instance | TextInstance | null>,
+  rootContainerInstance : Container,
+|};
+
+type TextInstance = {|
+  text : string,
+|};
+
+function removeOwner(element : React$Element) {
+  return {
+    ...element,
+    _owner: null,
+  }
+}
+
+const reconciler = ReactFiberReconciler({
+  getRootHostContext(rootContainerInstance : Container) {
+    return rootContainerInstance.context || emptyObject
+  },
+
+  getChildHostContext(parentHostContext, type) {
+    return parentHostContext;
+  },
+
+  createInstance(type : string, props : Props) : Instance {
+    return {
+      type,
+      props,
+      children: [],
+    };
+  },
+
+  appendInitialChild(parentInstance, child) {},
+  finalizeInitialChildren(parentInstance, type, props, rootContainerInstance) {},
+
+  prepareUpdate(instance, type, oldProps, newProps, hostContext) {},
+  commitUpdate(instance, type, oldProps, newProps, rootContainerInstance, internalInstanceHandle) {},
+  commitMount(instance, type, newProps, rootContainerInstance, internalInstanceHandle) {},
+
+  shouldSetTextContent(props) {},
+  resetTextContent(instance) {},
+
+  createTextInstance(text : string) : TextInstance {
+    return { text }
+  },
+  commitTextUpdate(textInstance, oldText, newText) {},
+
+  appendChild(parentInstance : Container | Instance, child : Instance) {
+    parentInstance.children.push({
+      $$typeof: child.$$typeof,
+      type: child.type,
+      key: child.key,
+      _owner: null,
+      _store: {},
+      ref: child.ref,
+      props: {
+        children: [...child.props.children].map(removeOwner),
+      },
+    });
+  },
+  insertBefore(parentInstance, child, beforeChild) {},
+  removeChild(parentInstance, child) {},
+
+  prepareForCommit() {},
+  resetAfterCommit(commitInfo) {},
+
+  scheduleAnimationCallback: window.requestAnimaionFrame,
+  scheduleDeferredCallback: window.requestIdleCallback,
+  useSyncScheduling: true,
+});
+
+class ReactShallowRendererFiber {
+  container = (null: ?Container);
+
+  render(element, context = emptyObject) {
+    this.container = {
+      context,
+      children: [],
+    };
+
+    const root = reconciler.createContainer(this.container);
+
+    reconciler.updateContainer(element, root, null, null);
+
+    return this.getRenderOutput();
+  }
+
+  getRenderOutput() {
+    const hasChildren = this.container.children.length > 0;
```

Comment:
Maybe too much info/review but:

* Composite components are the capitalized components like `<Foo />`
* Host components are the lowercase ones like `<div />`

I think what @gaearon is saying is that as a fiber renderer you don’t ever actually get to see the composite components because those are all happened internally so the `children` you see are the reconciled host component descriptions.